### PR TITLE
record_transformer: faster implementation of ruby placeholder expander

### PR DESCRIPTION
**Pull request by sonots**
revised ruby placeholder expander with 3x faster implementation

see benchmark https://gist.github.com/sonots/a1948ed576019653daa5

```
# Rehearsal --------------------------------------------------------
# placeholder            0.330000   0.010000   0.340000 (  0.330380)
# ruby placeholder       1.380000   0.020000   1.400000 (  1.422114)
# dynamic placeholder    0.530000   0.010000   0.540000 (  0.539724)
# new ruby placeholder   0.570000   0.000000   0.570000 (  0.576103)
# ----------------------------------------------- total: 2.850000sec
# 
#                            user     system      total        real
# placeholder            0.350000   0.010000   0.360000 (  0.355341)
# ruby placeholder       1.420000   0.020000   1.440000 (  1.449852)
# dynamic placeholder    0.510000   0.000000   0.510000 (  0.517271)
# new ruby placeholder   0.590000   0.010000   0.600000 (  0.591327)
```

The new ruby placeholder expander keeps backward compatibility and supports `record["key"]` placeholder newly. Also, `enable_ruby false` expander newly supports `record["key"]` placeholder.

### CODE REVIEW

File: lib/fluent/plugin/filter_record_transformer.rb [hunk 1]
```diff
@@ -175,17 +180,27 @@ def initialize(params)
...
+        placeholder_values.each do |key, value|
           if value.kind_of?(Array) # tag_parts, etc
             size = value.size
             value.each_with_index { |v, idx|
               placeholders.store("${#{key}[#{idx}]}", v)
               placeholders.store("${#{key}[#{idx-size}]}", v) # support [-1]
             }
+          elsif value.kind_of?(Hash) # record, etc
+            value.each {|k, v| placeholders.store("${#{k}}", v) } # foo
+            value.each {|k, v| placeholders.store(%Q[${#{key}["#{k}"]}], v) } # record["foo"]
```
Comment (repeatedly) [inline lib/fluent/plugin/filter_record_transformer.rb]: Why iterate 2 times? Following code doesn't work?

```rb
value.each { |k, v|
  placeholders.store("${#{k}}", v)
  placeholders.store(%Q[${#{key}["#{k}"]}], v)
}
```

Comment (sonots) [inline lib/fluent/plugin/filter_record_transformer.rb]: fixed

**Comment by repeatedly** [comment]
Please backport to v0.12

**Comment by sonots** [comment]
cherry-picked to v0.12 branch

**Pull request merged**